### PR TITLE
Support Parquet  Byte Stream Split Encoding

### LINKED ISSUE #4102: Parquet: Support `Encoding::BYTE_STREAM_SPLIT`
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
I would like to evaluate whether using the `BYTE_STREAM_SPLIT` encoding helps a Float64 column compress better. But it seems like it is not supported yet: https://github.com/apache/arrow-rs/blob/93484a10d145617434432d610e241640a06b382f/parquet/src/encodings/encoding/mod.rs#L90

**Describe the solution you'd like**
An implementation of the encoding. Even a naive, non-optimized version would resolve this issue. The implementation can be improved iteratively.

**Describe alternatives you've considered**
`PyArrow` seems to support it, but I would really like to stay within the Rust world.

**Additional context**
- Parquet format description here: https://github.com/apache/parquet-format/blob/master/Encodings.md#byte-stream-split-byte_stream_split--9
- The scalar impl in the C++ library is here: https://github.com/apache/arrow/blob/0bf777a5952be012e41f5b1ad443d4fec38e6f5a/cpp/src/arrow/util/byte_stream_split.h#L579-L602 . They also have SIMD variations, which will be more involved to port.

### PR DESCRIPTION
**Pull request by mwlon**
# Which issue does this PR close?

Closes #4102.

This builds on the previous attempt at a PR: #4183

# Rationale for this change
 
This brings us up to speed with the full set of Parquet encodings, I believe. It will also be important for the likely addition of f16 and fixed len byte arrays to the byte stream split encoding.

# What changes are included in this PR?

* implemented byte stream split encoding
* benchmark suite for encodings
  * Measured the performance as 3x faster than the previous PR's implementation. There are definitely more potential performance wins to be had with SIMD though.
* tests for the new encoding, including a test file in parquet-testing created via pyarrow.

# Are there any user-facing changes?

No API additions

### CODE REVIEW

**Comment by mwlon** [comment]
@tustvold

File: parquet/src/encodings/encoding/byte_stream_split_encoder.rs [hunk 1]
```diff
@@ -0,0 +1,76 @@
+use crate::basic::{Encoding, Type};
```
Comment (ggreco) [inline parquet/src/encodings/encoding/byte_stream_split_encoder.rs:1]: You need to add the apache license header to get it approved.

File: parquet/src/encodings/decoding/byte_stream_split_decoder.rs [hunk 2]
```diff
@@ -0,0 +1,104 @@
+use std::marker::PhantomData;
```
Comment (ggreco) [inline parquet/src/encodings/decoding/byte_stream_split_decoder.rs:1]: You need to add the apache license header to get it approved.

**Review by ggreco — APPROVED** [review]
I'm not part of the apache foundation but I will try to use your code ASAP in my application. I reviewed it and it seems good.

**Comment by pitrou** [comment]
> Measured the performance as 3x faster than the previous PR's implementation.

FTR, what the actual numbers?

File: parquet/src/arrow/arrow_reader/mod.rs [hunk 3]
```diff
@@ -1390,6 +1397,43 @@ mod tests {
...
+                .as_any()
+                .downcast_ref::<Float32Array>()
+                .unwrap();
+            let f64_col = batch
+                .column(1)
+                .as_any()
+                .downcast_ref::<Float64Array>()
+                .unwrap();
+
+            // This file contains floats from a standard normal distribution
```
Comment (pitrou) [inline parquet/src/arrow/arrow_reader/mod.rs:1424]: You can probably do an exact comparison on a few values as well, given that the file isn't going to change :-)

File: parquet/src/encodings/decoding.rs [hunk 4]
```diff
@@ -1760,6 +1786,29 @@ mod tests {
...
+    #[test]
+    fn test_byte_stream_split_multiple() {
+        let data = vec![
+            vec![
+                f32::from_le_bytes([0xAA, 0xBB, 0xCC, 0xDD]),
+                f32::from_le_bytes([0x00, 0x11, 0x22, 0x33]),
+            ],
+            vec![f32::from_le_bytes([0xA3, 0xB4, 0xC5, 0xD6])],
+        ];
+        test_byte_stream_split_decode::<FloatType>(data);
```
Comment (pitrou) [inline parquet/src/encodings/decoding.rs:1798]: Also add a test for DoubleType?

**Comment by mwlon** [comment]
> FTR, what are the actual numbers?

On my machine, it was about 17us for floats and 36us for doubles, pretty much the same speed for encoding and decoding. These are all just short of 4GB/s. The implementation I built from was around 1.2GB/s.

File: parquet/src/arrow/arrow_reader/mod.rs [hunk 5]
```diff
@@ -1390,6 +1397,43 @@ mod tests {
...
+            arrow::util::test_util::parquet_test_data(),
+        );
+        let file = File::open(path).unwrap();
+        let record_reader = ParquetRecordBatchReader::try_new(file, 128).unwrap();
+
+        let mut row_count = 0;
+        for batch in record_reader {
+            let batch = batch.unwrap();
+            row_count += batch.num_rows();
+            let f32_col = batch
```
Comment (tustvold) [inline parquet/src/arrow/arrow_reader/mod.rs:1413]: FWIW you can use https://docs.rs/arrow-array/latest/arrow_array/cast/trait.AsArray.html to make this less verbose

**Review by tustvold — APPROVED** [review]
The code makes sense to me, and looks well tested, thank you

**Comment by tustvold** [comment]
Perhaps we could update the README to no longer say that we don't support this encoding

File: parquet/src/arrow/arrow_reader/mod.rs (hunk 5 above)
Comment (mwlon) [reply parquet/src/arrow/arrow_reader/mod.rs]: 👍

File: parquet/src/encodings/decoding.rs (hunk 4 above)
Comment (mwlon) [reply parquet/src/encodings/decoding.rs:1787]: done 👍

**Comment by mwlon** [comment]
Updated the readme

**Comment by tustvold** [comment]
Thanks again

**Comment by ggreco** [comment]
@tustvold I see the V50 release is from 4 days ago, and I supposed it should contain this change, but I tried to update to v50 and enable BYTE_STREAM_SPLIT and I get this error:
```
thread 'upload::tests::test_upload_sensor' panicked at 'called `Result::unwrap()` on an `Err` value: NYI("Encoding BYTE_STREAM_SPLIT is not supported")', /Users/gabry/.cargo/registry/src/index.crates.io-6f17d22bba15001f/parquet-50.0.0/src/column/writer/mod.rs:249:58
```

**Comment by tustvold** [comment]
The release was cut prior to this being merged, as it goes throuh an ASF mandated RC process that takes at least 3 days. You can view the changelog and/or the git history to see what made a particular release.

**Comment by alamb** [comment]
Tracking next release in https://github.com/apache/arrow-rs/issues/5453

**Pull request merged**